test/e2e/framework/framework.go

```diff
@@ -0,0 +1,295 @@
+    ginkgo.AfterEach(f.AfterEach)
+
+    return f
+}
+
+
+func (f *Framework) BeforeEach() {
+    // workaround for a bug in ginkgo.
+    // https://github.com/onsi/ginkgo/issues/222
+    f.cleanupHandle = AddCleanupAction(f.AfterEach)
+
+    ginkgo.By("Creating kubernetes clients")
+    for _, context := range TestContext.KubeContexts {
+        client := f.createKubernetesClient(context)
+        f.ClusterClients = append(f.ClusterClients, client)
+    }
+
+    if !f.SkipNamespaceCreation {
+        ginkgo.By(fmt.Sprintf("Building namespace api objects, basename %s", f.BaseName))
+
+        namespaceLabels := map[string]string {
+            "e2e-framework": f.BaseName,
+        }
+
+        for idx, clientSet := range f.ClusterClients {
+            switch idx {
+            case ClusterA: // On the first cluster we let k8s generate a name for the namespace
+                namespace := generateNamespace(clientSet, f.BaseName, namespaceLabels)
+                f.Namespace = namespace.GetName()
+                f.UniqueName = namespace.GetName()
+            default: // On the other clusters we use the same name to make tracing easier
+                f.CreateNamespace(clientSet, f.Namespace, namespaceLabels)
+            }
+        }
+    } else {
+        f.UniqueName = string(uuid.NewUUID())
+    }
+
+}
+
+func (f *Framework) createKubernetesClient(context string) *kubeclientset.Clientset {
+
+    restConfig, _, err := loadConfig(TestContext.KubeConfig, context)
+    if err != nil {
+        Errorf("Unable to load kubeconfig file %s for context %s, this is a non-recoverable error",
+            TestContext.KubeConfig, context)
+        Errorf("loadConfig err: %s", err.Error())
+        os.Exit(1)
+    }
+
+    testDesc := ginkgo.CurrentGinkgoTestDescription()
+    if len(testDesc.ComponentTexts) > 0 {
+        componentTexts := strings.Join(testDesc.ComponentTexts, " ")
+        restConfig.UserAgent = fmt.Sprintf(
+            "%v -- %v",
+            rest.DefaultKubernetesUserAgent(),
+            componentTexts)
+    }
+
+    restConfig.QPS = f.Options.ClientQPS
+    restConfig.Burst = f.Options.ClientBurst
+    if f.Options.GroupVersion != nil {
+        restConfig.GroupVersion = f.Options.GroupVersion
+    }
+    clientSet, err := kubeclientset.NewForConfig(restConfig)
+    Expect(err).NotTo(HaveOccurred())
+
+    // create scales getter, set GroupVersion and NegotiatedSerializer to default values
+    // as they are required when creating a REST client.
+    if restConfig.GroupVersion == nil {
+        restConfig.GroupVersion = &schema.GroupVersion{}
+    }
+    if restConfig.NegotiatedSerializer == nil {
+        restConfig.NegotiatedSerializer = scheme.Codecs
+    }
+    return clientSet
+}
+
+
+func deleteNamespace(client kubeclientset.Interface, namespaceName string) error {
+
+    err := client.CoreV1().Namespaces().Delete(
+        namespaceName,
+        &metav1.DeleteOptions{});
+
+    return err
```

Comment:
This could be shortened to:

    return client.CoreV1()...

Reply:
true 👍 it was more complicated at some point, but now that's not very useful